src/utils/cache.rs

```diff
@@ -62,20 +62,65 @@ pub trait CacheExt: Cache {
         client: &reqwest::blocking::Client,
         url: &String,
         timeout: usize,
-    ) -> Result<String> {
-        let data: String =
-            match self.fetch(&url) {
-                Some(cached) => cached,
-                None => {
-                    let response = client.get(url).send()?;
-                    // Don't cache if it is a Server error
-                    if response.status().is_server_error() { anyhow::bail!("Got server error for {}", url); };
-                    let raw_data = response.text()?;
-                    self.create(&url, &raw_data, timeout);
-                    raw_data
+    ) -> ApiResult<String> {
+        match self.fetch(&url) {
+            Some(cached) => {
+                let cached_with_code = CachedWithCode::split(&cached);
+                if !cached_with_code.is_error() {
+                    Ok(String::from(cached_with_code.data))
+                } else {
+                    return if let Ok(backend_error) = serde_json::from_str::<BackendError>(&cached_with_code.data) {
+                        Err(ApiError { status: cached_with_code.code, message: ApiErrorMessage::BackendError(backend_error) })
+                    } else {
+                        Err(ApiError { status: cached_with_code.code, message: ApiErrorMessage::SingleLine(cached_with_code.data) })
+                    };
                 }
-            };
-        Ok(data)
+            }
+            None => {
+                let response = client.get(url).send()?;
+                if response.status().is_server_error() {
+                    return Err(anyhow::anyhow!("Got server error for {}", url).into());
```

Comment:
Why not use an ApiError here?

Reply:
I was too lazy to initialise `status_code` earlier xD